File: Data Structures & Algorithms/sliding-window-maximum/submission-2.cs
public class Solution {
    // Using deque (Double-ended queue)
    public int[] MaxSlidingWindow(int[] nums, int k) {
       if (nums.Length == 0) return [];

        int[] res = new int[nums.Length - k + 1];
        LinkedList<int> deque = new();

        // Pre-fill the first window
        for (int i = 0; i < k; i++)
        {
            while (deque.Count > 0 && nums[deque.Last!.Value] < nums[i])
                deque.RemoveLast();
            deque.AddLast(i);
        }

        for (int l = 0; l <= nums.Length - k; l++)
        {
            int r = l + k - 1;

            // Record max for this window
            res[l] = nums[deque.First!.Value];

            // Remove l if it's the front (it's leaving the window)
            if (deque.First!.Value == l)
                deque.RemoveFirst();

            // Add the next element if there is one
            if (r + 1 < nums.Length)
            {
                while (deque.Count > 0 && nums[deque.Last!.Value] < nums[r + 1])
                    deque.RemoveLast();
                deque.AddLast(r + 1);
            }
        }

        return res;
    }
}

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-2.go
func lengthOfLIS(nums []int) int {
	n := len(nums)
    dp := make([]int, n)

	lis := 0
	for i := 0; i < n; i++ {
		dp[i] = 1
		for j := 0; j < i; j++ {
			if nums[i] > nums[j] {
				dp[i] = max(dp[i], dp[j] + 1)
			}
		}
		lis = max(lis, dp[i])
	}
	return lis
}
